enhancements/declarative-index-config.md

```diff
@@ -576,141 +618,140 @@ $ tree community-operators
-            {
+    "channels": ["alpha"],
+    "description": "A message about etcd operator, a description of channels"
+}
+{
+    "schema": "olm.bundle",
+    "name": "etcdoperator-community.v0.6.1",
+    "package": "etcd",
+    "image": "quay.io/operatorhubio/etcd:v0.6.1",
+    "version": "0.6.1",
+    "properties":[
+        {
+            "type": "olm.package.provided",
+            "value": {
+                "packageName": "etcd",
+                "version": "0.6.1"
+            }
+        },
+        {
+            "type": "olm.gvk.provided",
+            "value": {
+                "group": "etcd.database.coreos.com",
+                "kind": "EtcdCluster",
+                "version": "v1beta2"
+            }
+        },
+        {
+            "type": "olm.channel",
+            "value": {
+                "name": "alpha"
+            }
+        }
+    ],
+    "relatedImages": [
+        {
             "name": "etcdv0.6.1",
             "image": "quay.io/coreos/etcd-operator@sha256:bd944a211eaf8f31da5e6d69e8541e7cada8f16a9f7a5a570b22478997819943"
-            }
-        ]
-    }
-]
+        }
+    ]
+}
 $ opm index add --bundles quay.io/operatorhubio/etcd:v0.9.0 --mode replaces --tag docker.io/my-namespace/community-operators:latest
 $ docker push docker.io/my-namespace/community-operators:latest
 $ opm index inspect --image=docker.io/my-namespace/community-operators:latest --output=json
 $ cat community-operators/etcd.json
-[
-    {
-        "schema": "olm.package",
-        "name": "etcd",
-        "defaultChannel": "singlenamespace-alpha",
-        "icon": {
-            "base64data":"iVBORw0KGgoAAAANSUhEUgAAA.....",
-            "mediatype":"image/png"
-        },
-        "channels": ["alpha", "singlenamespace-alpha"],
-        "description": "A message about etcd operator, a description of channels"
+{
+    "schema": "olm.package",
+    "name": "etcd",
+    "defaultChannel": "singlenamespace-alpha",
+    "icon": {
+        "base64data":"iVBORw0KGgoAAAANSUhEUgAAA.....",
+        "mediatype":"image/png"
     },
-    {
-        "schema": "olm.bundle", 
-        "name": "etcdoperator-community.v0.6.1",
-        "package": "etcd",
-        "image": "quay.io/operatorhubio/etcd:v0.6.1",
-        "version": "0.6.1",
-        "properties":[
-            {
-                "name": "olm.package",
-                "values": {
-                    "packageName": "etcd",
-                    "version": "0.6.1"
-                }
-            },
-            {
-                "name":"olm.gvk",
-                "type": "provided",
-                "values":{
-                    "group": "etcd.database.coreos.com", 
-                    "kind": "EtcdCluster", 
-                    "version": "v1beta2"
-                }
-            },
-            {
-                "name": "olm.channel",
-                "value": "alpha",
-            }
-        ],
-        "relatedImages": [
-            {
+    "channels": ["alpha", "singlenamespace-alpha"],
```

Comment:
I still don't think we should have this at the package level.  This information can(and should) be constructed from the channel information already included in the list of bundles.  Consumers of the json can cache that construction as needed, but duplicating the information in the json itself just makes it harder to maintain the file and more likely to make an error.

Reply:
> Would we be able to detect/prevent that?

that's fair, no we would not.  In that sense you are intending to treat this list as "valid channels for the package" (is it ok to list a channel here that no bundle belongs to?).  

I can see value in that, but i'm still weighing it against the cost for someone who wants to build an override mechanism(namely the operator pipeline) that mutates bundles in place (i.e. moves a bundle from one channel to another or adds additional channels a bundle belongs to).  They will have to reconcile the package level channel list to accomplish that, which, while certainly not impossible or all that difficult, it is an extra burden.

At a minimum if we're going to pitch it as a validation mechanism, i'd want to change the field name to reflect its intent.  e.g. validChannelNames or something.

Reply:
That seems alright to me. We could also make it optional and only validate channels listed in bundles if the package's `validChannelNames` field is present and non-empty.

Reply:
thinking about this some more, it's still awkward since `opm add` isn't going to treat that list as a validation.  Presumably if i add a new bundle that specifies a new channel, `opm add` is going to have to accept that new channel name as valid and add it to the list of channels here.  Which means if someone adds a bundle in the "v1.0" channel instead of the "1.0" channel, we'll still end up with both channels and the error won't be caught.

So the only time "validation" can be done is when someone is hand-editing the package json or using some other tool to modify the package json.  But when opm add is used, we trust you not to screw up the channel names in your bundle.  Is that useful?

Reply:
I'm not sure that has to be true, does it? I think `opm index/registry add` could use it if present.

By default, that field would be unset or empty, so `opm index add` would work as you describe. It would blindly trust the channel properties from any added bundles.

But if a package maintainer wants to get that extra validation, they could hand-edit the `olm.package`, set that field, and from then on the `add` commands would validate any newly added bundles. If a package maintainer actually wants to add a new channel, they would just need to hand-edit again to add the new channel to the list prior to running the `add` command.

Reply:
> By default, that field would be unset or empty, so opm index add would work as you describe. It would blindly trust the channel properties from any added bundles.

got it, sorry i forgot the plan was to have empty mean "accept anything" and also that opm add wouldn't update the field, it would only be updated manually to specify the valid list.

Reply:
Ended up removing this field per https://github.com/operator-framework/enhancements/pull/69#discussion_r592071833